### ISSUE TITLE
Amesos2: Linear system segfaults w/ SuperLU, SuperLU_DIST, & UMFPACK, but not with KLU2

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#3647]
@trilinos/amesos2 

@vbrunini found a Tpetra linear system on a single MPI process, that makes Amesos2 segfault when using `SuperLU`, `SuperLU_DIST`, or `UMFPACK`.  However, `KLU2` does not segfault.

I have put the linear system through R&A and will post it as a pull request.  I just want to make this issue now so I can put the issue number in the PR.

@vbrunini 's guess is that Amesos2 does not like matrices that have "gaps" in their row Map.  That is the case for the matrix in the test that I will post.

**Comment by srajama1** [comment]
KLU2 takes a special path for single rank cases compared to other codes. The bug might be in the code that we don't use for KLU2. @ndellingwood wanted to propagate these KLU2 changes to other solvers anyway. This might be the time to do it.

**Comment by vbrunini** [comment]
Since you mentioned a special code path for single rank cases I just tried running the problem this matrix is from on 2 ranks and in that case KLU2 also segfaults.

**Comment by srajama1** [comment]
Thanks for confirming it. That helps. We will resolve it. Is this blocking you from making progress ? I am asking to understand the timeline. We are busy with a Kokkos integration into Trilinos, we can look at this after that is done. Is that reasonable ?

**Comment by vbrunini** [comment]
Yeah no problem.

**Comment by srajama1** [comment]
@ndellingwood is looking at this

**Comment by vbrunini** [comment]
Any update on this?

**Comment by ndellingwood** [comment]
@vbrunini working on it this week, sorry for the delay I've been a bit overloaded the past few weeks but aiming to put in a PR with a fix this week.
Brief update - some previous work to handle non-contiguous GIDs was intended to resolve cases like this, but this case exposed some shortcomings I'm working to resolve. The single-thread run on KLU2 happened to run to completion but the results are not correct (confirmed by testing the matrix Mark provided with a different RHS and checking the residual). When setting a parameter to expect non-contiguous GIDs the code seg-faults, which exposed the existing issue. I've created a much smaller (5x5) single node reproducer that I'm using to work through the issue, will modify it to use the provided matrix and rhs as a test in Amesos2 once resolved.

**Comment by vbrunini** [comment]
Great, thanks. Once you've got something working I can test it out against more of our test suite to make sure there isn't anything else unusual we are doing with Maps that isn't captured by this one test matrix.

**Comment by ndellingwood** [comment]
@vbrunini I put in PR #4454 with a fix that worked with my single-proc example tested with KLU2. PR is WIP until I finish cleaning the test and check with other solvers, but if you'd like to test please pull that PR (single line change) and add the parameter list option in the comment there to indicate to Amesos2 your matrix has non-contigous (i.e. gapped) rowids.

**Comment by vbrunini** [comment]
Thanks, I'll give this a shot today.

**Comment by vbrunini** [comment]
@ndellingwood I pulled your fix and am still seeing issues. Just noticed that you said you tested with KLU2, but when we initially put this in KLU2 worked, but SuperLU, SuperLU_dist, and UMFPACK segfaulted. That is still consistent with what I see with your patch applied.

**Comment by ndellingwood** [comment]
@vbrunini thanks for the fast response, a couple questions:

1. Just to confirm, did you set the parameter for each solver to indicate the matrix rowmap is non-contiguous? This needs to happen before the solver is created
e.g.
```c++
std::string solverName = "KLU2"; //or SuperLU, UMFPACK, etc
Teuchos::ParameterList amesos2_params("Amesos2");
amesos2_params.sublist(solverName).set("IsContiguous", false, "Are GIDs Contiguous");
```

2. Did the KLU2 results change? Even though KLU2 ran prior with a single rank, the results were likely incorrect. I'm curious if either the PR change or setting the parameter that the rowmap is non-contiguous affected the results. 

In my test case with a smaller matrix with gapped rowids, prior to the PR I could run the test on a single proc but yielding incorrect results; the run would seg fault when setting `"IsContiguous", false`. After the PR change the code gives correct results when setting the parameter for `"IsContiguous", false`. I'll begin testing with the other TPLs to chase down the issue you're seeing and expand the test to handle the badmatrix provided here.

**Comment by vbrunini** [comment]
1. Yes, or at least I have attempted to. Is there a way to get Amesos2 to report back what parameters it is actually using to make sure I didn't screw up attempting to set it?

2. I'm seeing the same results before/after with KLU2.

**Comment by ndellingwood** [comment]
@vbrunini thanks for confirming, there is no getter in Amesos2 to return a parameter list, makes sense to put one in. The issue I put in the PR for isnt't the same one you are running into, but I'll continue to add to #4454 once I get this fixed.

**Comment by ndellingwood** [comment]
PRs #4454 and #4527 should resolve the issue with KLU2, SuperLU and Umfpack. Will follow up regarding SuperLU_Dist.

**Comment by ndellingwood** [comment]
@vbrunini PRs #4454 and #4527 are now merged into the develop branch, please let us know if any issues persist.

**Comment by mhoemmen** [comment]
@ndellingwood

**Comment by vbrunini** [comment]
I think this is good to close.
